edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs

```diff
@@ -0,0 +1,103 @@
+        readonly AsyncLock cloudProxyLock = new AsyncLock();
+        readonly Func<Task<Try<ICloudProxy>>> cloudProxyGetter;
+
+        ICloudProxy innerCloudProxy;
+
+        public RetryingCloudProxy(Func<Task<Try<ICloudProxy>>> cloudProxyGetter, ICloudProxy cloudProxyImplementation)
+        {
+            this.cloudProxyGetter = Preconditions.CheckNotNull(cloudProxyGetter, nameof(cloudProxyGetter));
+            this.innerCloudProxy = Preconditions.CheckNotNull(cloudProxyImplementation, nameof(cloudProxyImplementation));
+        }
+
+        public bool IsActive => this.innerCloudProxy.IsActive;
+
+        public Task<bool> CloseAsync() => this.ExecuteOperation(c => c.CloseAsync());
+
+        public Task<bool> OpenAsync() => this.ExecuteOperation(c => c.OpenAsync());
+
+        public Task SendMessageAsync(IMessage message) => this.ExecuteOperation(c => c.SendMessageAsync(message));
+
+        public Task SendMessageBatchAsync(IEnumerable<IMessage> inputMessages) => this.ExecuteOperation(c => c.SendMessageBatchAsync(inputMessages));
+
+        public Task UpdateReportedPropertiesAsync(IMessage reportedPropertiesMessage) => this.ExecuteOperation(c => c.UpdateReportedPropertiesAsync(reportedPropertiesMessage));
+
+        public Task<IMessage> GetTwinAsync() => this.ExecuteOperation(c => c.GetTwinAsync());
+
+        public Task SendFeedbackMessageAsync(string messageId, FeedbackStatus feedbackStatus) => this.ExecuteOperation(c => c.SendFeedbackMessageAsync(messageId, feedbackStatus));
+
+        public Task SetupCallMethodAsync() => this.ExecuteOperation(c => c.SetupCallMethodAsync());
+
+        public Task RemoveCallMethodAsync() => this.ExecuteOperation(c => c.RemoveCallMethodAsync());
+
+        public Task SetupDesiredPropertyUpdatesAsync() => this.ExecuteOperation(c => c.SetupDesiredPropertyUpdatesAsync());
+
+        public Task RemoveDesiredPropertyUpdatesAsync() => this.ExecuteOperation(c => c.RemoveDesiredPropertyUpdatesAsync());
+
+        public Task StartListening() => this.ExecuteOperation(c => c.StartListening());
+
+        internal ICloudProxy InnerCloudProxy => this.innerCloudProxy;
+
+        Task ExecuteOperation(Func<ICloudProxy, Task> func) => this.ExecuteOperation(
+            async c =>
+            {
+                await func(c);
+                return 1;
+            });
+
+        async Task<T> ExecuteOperation<T>(Func<ICloudProxy, Task<T>> func)
+        {
+            for (int i = 0; i < RetryCount; i++)
+            {
+                ICloudProxy cloudProxy = await this.GetCloudProxy();
+                try
+                {
+                    return await func(cloudProxy);
+                }
+                catch (Exception)
+                {
+                    if (cloudProxy.IsActive || i + 1 == 3)
+                    {
+                        throw;
+                    }
+                }
+            }
+
+            // Should never get here
+            return default(T);
```

Comment:
I think the compiler will allow you to remove this unreachable code if you change the `for` loop condition from `i < RetryCount` to `true`. Your `catch` clause already controls when you will exit the loop, so you don't need it twice. When you change it to `true`, the compiler knows that the _only_ way to exit the function is from inside the loop.

Reply:
Good point, fixed.